File: web/src/lib/syncMeta.ts
// Module-level sync coordination state. Deliberately NOT persisted: it only
// describes network operations in flight during the current page session.

// itemId -> number of writes in flight for that record.
const pendingWrites = new Map<string, number>();

export function beginWrite(id: string) {
    pendingWrites.set(id, (pendingWrites.get(id) || 0) + 1);
}

export function endWrite(id: string) {
    const count = pendingWrites.get(id) || 0;
    if (count <= 1) pendingWrites.delete(id);
    else pendingWrites.set(id, count - 1);
}

export function hasPendingWrite(id: string): boolean {
    return pendingWrites.has(id);
}

// Temp ("local_") ids whose create request is currently in flight, so the
// reconnect push doesn't create them a second time.
export const pendingCreates = new Set<string>();

export function parsePbDate(s: string | undefined | null): number {
    if (!s) return 0;
    const t = Date.parse(s);
    return Number.isNaN(t) ? 0 : t;
}

// A realtime update event must be ignored when we have our own write in
// flight for the record (our optimistic state is newer and the in-flight
// write will supersede the event) or when the event's server timestamp is
// not newer than the version we already applied (stale/out-of-order event).
export function shouldSkipRemoteUpdate(
    id: string,
    recordUpdated: string | undefined,
    knownServerUpdated: number | undefined
): boolean {
    if (hasPendingWrite(id)) return true;
    const incoming = parsePbDate(recordUpdated);
    return incoming > 0 && knownServerUpdated !== undefined && incoming <= knownServerUpdated;
}

File: android/app/src/main/java/com/bor_devs/shoplist/data/sync/SyncMeta.kt
package com.bor_devs.shoplist.data.sync

import java.time.Instant
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Module-level sync coordination, ported from web/src/lib/syncMeta.ts. Tracks
 * writes in flight per record and decides when an incoming realtime update is
 * a stale echo that must be ignored. Deliberately not persisted.
 */
@Singleton
class SyncMeta @Inject constructor() {
    private val pendingWrites = ConcurrentHashMap<String, Int>()

    /** Temp ("local_") ids whose create request is currently in flight. */
    val pendingCreates: MutableSet<String> = ConcurrentHashMap.newKeySet()

    fun beginWrite(id: String) {
        pendingWrites.merge(id, 1, Int::plus)
    }

    fun endWrite(id: String) {
        pendingWrites.compute(id) { _, count -> if ((count ?: 0) <= 1) null else count!! - 1 }
    }

    fun hasPendingWrite(id: String): Boolean = pendingWrites.containsKey(id)

    /**
     * Ignore a realtime update when we have our own write in flight for the
     * record, or when the event's server timestamp is not newer than the
     * version we already applied (stale / out-of-order).
     */
    fun shouldSkipRemoteUpdate(id: String, recordUpdated: String?, knownServerUpdated: Long): Boolean {
        if (hasPendingWrite(id)) return true
        val incoming = parsePbDate(recordUpdated)
        return incoming > 0 && knownServerUpdated > 0 && incoming <= knownServerUpdated
    }

    companion object {
        fun parsePbDate(s: String?): Long {
            if (s.isNullOrBlank()) return 0
            return try {
                // PocketBase format: "2024-01-15 14:30:00.123Z" -> ISO instant.
                val iso = s.trim().replace(' ', 'T')
                Instant.parse(iso).toEpochMilli()
            } catch (_: Exception) {
                0
            }
        }
    }
}
